codec-http/src/main/java/io/netty/handler/codec/http/HttpObjectDecoder.java

```diff
@@ -1113,8 +1113,14 @@ private static int findNonWhitespace(byte[] sb, int offset, int end) {
 
     private static int findEndOfString(byte[] sb, int start, int end) {
         for (int result = end - 1; result > start; --result) {
-            if (!isWhitespace(sb[result])) {
+            byte c = sb[result];
+            if (!isWhitespace(c)) {
                 return result + 1;
+            } else if (!isOWS(c)) {
+                // Only OWS is supported for whitespace
+                // See https://datatracker.ietf.org/doc/html/rfc9110#section-5.5
+                throw new IllegalArgumentException("Invalid end of value, only a space or horizontal tab allowed," +
+                        " but received a '" + c + "' (0x" + Integer.toHexString(c) + ")");
```

Comment:
A thought: the default header validator will already reject these chars if they're allowed to get there (along with NULL which falls into the same class of rejected chars but wont be caught here) and if someone wants to accept anything they can override the validator. However, this may open the door to other problems such as request smuggling so I think your suggestion might be the safest way forward.

```suggestion
            if (!isOWS(sb[result])) {
                return result + 1;
            }
```